File: benchmarks/eisenstein-norm/eisenstein_norm.c
/* Eisenstein norm benchmark — C.
 *
 * Eisenstein norm: |a + bω|² = a² + ab + b²
 * Computes 1,000,000 norms for pairs (a,b) cycling through 0..999.
 */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <assert.h>

#define ITERS 1000000
#define RANGE 1000

static inline long long eisenstein_norm(int a, int b) {
    long long la = (long long)a;
    long long lb = (long long)b;
    return la * la + la * lb + lb * lb;
}

int main(void) {
    /* Generate input pairs */
    int *a_arr = malloc(ITERS * sizeof(int));
    int *b_arr = malloc(ITERS * sizeof(int));
    if (!a_arr || !b_arr) {
        fprintf(stderr, "malloc failed\n");
        return 1;
    }

    for (int i = 0; i < ITERS; i++) {
        a_arr[i] = i % RANGE;
        b_arr[i] = (i / RANGE) % RANGE;
    }

    /* Verify known values */
    assert(eisenstein_norm(0, 0) == 0);
    assert(eisenstein_norm(1, 0) == 1);
    assert(eisenstein_norm(1, 1) == 3);
    assert(eisenstein_norm(3, 5) == 49);
    assert(eisenstein_norm(7, 3) == 79);

    /* Benchmark */
    struct timespec start, end;
    clock_gettime(CLOCK_MONOTONIC, &start);

    long long total = 0;
    for (int i = 0; i < ITERS; i++) {
        total += eisenstein_norm(a_arr[i], b_arr[i]);
    }

    clock_gettime(CLOCK_MONOTONIC, &end);
    double elapsed = (end.tv_sec - start.tv_sec) + (end.tv_nsec - start.tv_nsec) / 1e9;
    double throughput = ITERS / elapsed;

    printf("C Eisenstein Norm Benchmark\n");
    printf("============================\n");
    printf("Iterations:  %d\n", ITERS);
    printf("Time:        %.4fs\n", elapsed);
    printf("Throughput:  %.0f ops/sec\n", throughput);
    printf("Checksum:    %lld\n", total);
    printf("\n");

    free(a_arr);
    free(b_arr);
    return 0;
}

File: benchmarks/eisenstein-norm/eisenstein_norm.rs
/* Eisenstein norm benchmark — Rust.
 *
 * Eisenstein norm: |a + bω|² = a² + ab + b²
 * Computes 1,000,000 norms for pairs (a,b) cycling through 0..999.
 */

use std::time::Instant;

const ITERS: u64 = 1_000_000;
const RANGE: i64 = 1000;

#[inline(always)]
fn eisenstein_norm(a: i64, b: i64) -> i64 {
    a * a + a * b + b * b
}

fn main() {
    // Generate input pairs
    let mut a_arr = Vec::with_capacity(ITERS as usize);
    let mut b_arr = Vec::with_capacity(ITERS as usize);
    for i in 0..ITERS {
        a_arr.push((i % RANGE as u64) as i64);
        b_arr.push(((i / RANGE as u64) % RANGE as u64) as i64);
    }

    // Verify known values
    assert_eq!(eisenstein_norm(0, 0), 0);
    assert_eq!(eisenstein_norm(1, 0), 1);
    assert_eq!(eisenstein_norm(1, 1), 3);
    assert_eq!(eisenstein_norm(3, 5), 49);
    assert_eq!(eisenstein_norm(7, 3), 79);

    // Benchmark
    let start = Instant::now();
    let mut total: i64 = 0;
    for i in 0..ITERS as usize {
        total += eisenstein_norm(a_arr[i], b_arr[i]);
    }
    let elapsed = start.elapsed();
    let elapsed_s = elapsed.as_secs_f64();
    let throughput = ITERS as f64 / elapsed_s;

    println!("Rust Eisenstein Norm Benchmark");
    println!("================================");
    println!("Iterations:  {}", ITERS);
    println!("Time:        {:.4}s", elapsed_s);
    println!("Throughput:  {:.0} ops/sec", throughput);
    println!("Checksum:    {}", total);
    println!();
}
